src/core/peliasgeocoder.cpp

```diff
@@ -0,0 +1,209 @@
+/***************************************************************************
+  peliasgeocoder.cpp
+  ---------------
+  Date                 : March 2021
+  Copyright            : (C) 2021 by Mathieu Pellerin
+  Email                : mathieu at opengis dot ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "peliasgeocoder.h"
+#include "qgsblockingnetworkrequest.h"
+#include "qgsgeocodercontext.h"
+#include "qgslogger.h"
+#include "qgsnetworkaccessmanager.h"
+
+#include <QDateTime>
+#include <QUrl>
+#include <QUrlQuery>
+#include <QMutex>
+#include <QNetworkRequest>
+#include <QJsonDocument>
+#include <QJsonArray>
+
+QMutex PeliasGeocoder::sMutex;
+QMap< QUrl, QList< QgsGeocoderResult > > PeliasGeocoder::sCachedResults;
```

Comment:
Warning - you should use qglobalstatic on this or you'll get crashes on exit

Reply:
@nyalldawson , thanks for the tip, adjusted.